devicetypes/plaidsystems/spruce-controller.src/spruce-controller.groovy

```diff
@@ -14,736 +10,570 @@
-    String onoff
-    if(descMap.value == '00'){
-    	onoff = 'off'        
-    }    
-    else onoff = 'on'
-            
-    if (EP == 1){
-    	map.name = 'switch'
-        map.value = onoff
-        map.descriptionText = "${device.displayName} turned sprinkler program ${onoff}"
-        }
-        
-    else if (EP == 18) {
-        map.name = 'rainsensor'
-    	log.debug "Rain enable: ${RainEnable}, sensor: ${onoff}"
-        map.value = 'rainSensor' + onoff
-        map.descriptionText = "${device.displayName} rain sensor is ${onoff}"
-        }
-   	else {
-        EP -= 1
-        map.name = 'switch' + EP
-    	map.value = 'z' + EP + onoff
-    	map.descriptionText = "${device.displayName} turned Zone $EP ${onoff}"
-    	}
-        
-    map.isStateChange = true 
-    map.displayed = true    
-    return map
-}
-
-def getAlarm(descMap){
-	def map = [:]
-    map.name = 'status'
-    def alarmID = Integer.parseInt(descMap.value.trim(), 16)
-    log.debug "${alarmID}"
-    map.value = 'alarm'
-    map.displayed = true
-    map.isStateChange = true
-    if(alarmID <= 0){
-    	map.descriptionText = "${device.displayName} reboot, no other alarms"
-        map.value = 'warning'
-        //map.isStateChange = false
-        }
-    else map.descriptionText = "${device.displayName} reboot, reported zone ${alarmID - 1} error, please check zone is working correctly, press SYNC SETTINGS button to clear"
-       
-    return map        
-}
-
-//status notify and change status
-def notify(String val, String txt){
-	sendEvent(name: 'status', value: val, descriptionText: txt, isStateChange: true, display: false)
-    
-    //String txtShort = txt.take(100)
-    sendEvent(name: 'tileMessage', value: txt, descriptionText: "", isStateChange: true, display: false)
-}
-
-def updated(){
-	log.debug "updated"
-    
+def commandType(endpoint, cluster){
+	if (cluster == 9) return "alarm"
+	else if (endpoint == 1) return "schedule"
+	else if (endpoint in 2..17) return "zone"
+	else if (endpoint == 18) return "rainsensor"
+	else if (endpoint == 19) return "refresh"
 }
 
-//prefrences - rain sensor, manual time
-def rain() {
-    log.debug "Rain sensor: ${RainEnable}"
-    if (RainEnable) sendEvent(name: 'rainsensor', value: 'enable', descriptionText: "${device.displayName} rain sensor is enabled", isStateChange: true)
-    else sendEvent(name: 'rainsensor', value: 'disable', descriptionText: "${device.displayName} rain sensor is disabled", isStateChange: true)
-    
-    if (RainEnable) "st wattr 0x${device.deviceNetworkId} 18 0x0F 0x51 0x10 {01}"
-    else "st wattr 0x${device.deviceNetworkId} 18 0x0F 0x51 0x10 {00}"
+//--------------------end zigbee parse-------------------------------//
+
+def installed() {	
+    //configure() called after installed()
 }
 
-def manualTime(value){	
-	sendEvent(name: 'minutes', value: "${value}", descriptionText: "Manual Time set to ${value}", display: false)
+def updated() {
+	log.debug "updated"
+
+	createChildDevices()
```

Comment:
can you wrap this in an

`if (!childDevices)` just to avoid possible messy behavior from overwrite attempts?

Reply:
He avoids it here https://github.com/SmartThingsCommunity/SmartThingsPublic/pull/47339/files#diff-203530982adf372057d9eaac68564e740dc55e3cc174c90d1d60ad524ec56834R227